### ISSUE TITLE
Find an element by presence of some attribute, not its value

### ISSUE THREAD

**Issue by nonkor** [watir/watir#345]
Sometimes there is needed to find element by the presence attribute, not its value. For example, on our project we have elements with different value of `data-tooltip` attribute: 

``` slim
.some_element data={tooltip: 'blah-blah'}
```

And we need to check this tooltip in last step our test. Currently we use css selector to search element with tooltip: 

``` css
span[data-tooltip]
```

It would be great to have possibility to find element by watir way, so we could use common accessor for it like

``` ruby
browser.span(data_tooltip: true)
```

It could be other syntax solution, doesn't matter. What do you think about this improvement?

**Comment by p0deje** [comment]
I'm personally fine with this change with a couple of concerns:

1. It will require plenty of element_locator refactoring.
2. I think the API should be something like:

```ruby
browser.span(attribute: 'data-tooltip')
```

**Comment by jkotests** [comment]
Assuming that implementing this would also address the ability to locate elements via boolean attributes, I would vote to go with @nonkor proposed API:

````````
browser.span(data_tooltip: true)
````````

A couple of reasons:
* It is easier to read when there are multiple attributes - eg a checked and disabled checkbox, `browser.checkbox(checked: true, disabled: true)`
* It would allow for specifying the presence of, as well as lack of presence - eg an unchecked and disabled checkbox, `browser.checkbox(checked: false, disabled: true)`
* It seems to better follow the [HTML specification for boolean attributes](http://www.w3.org/html/wg/drafts/html/master/infrastructure.html#boolean-attributes), which says "The presence of a boolean attribute on an element represents the true value, and the absence of the attribute represents the false value."

**Comment by p0deje** [comment]
It makes sense to me, however, in this case we'll need to somehow limit the possible boolean attributes to the ones that are generated from HTML spec (e.g. [input](https://github.com/watir/watir-webdriver/blob/master/lib/watir-webdriver/elements/generated.rb#L447)).

On the other hand, there is a bit of confusion for non-boolean attributes because `data_tooltip: true` maybe be treated as `[data-tooltip="true"]`.

**Comment by jkotests** [comment]
I was thinking that a boolean specifier value would simply denote the presence of the attribute. It could be applied to any attribute, regardless of whether or not it is actually an HTML boolean attribute. The existing behaviour for String and Regexp specifier values would remain unchanged - ie would continue to check the attribute value.

For example, if the HTML was:

````````
<div data-tooltip="false">text</div>
````````

Then using a String specifier value would continue to check the attribute value:

````````
browser.div(data_tooltip: "true").present?
#=> false

browser.div(data_tooltip: "false").present?
#=> true
````````

If you use a Boolean specifier value, it just checks the presence of the attribute (it does not care what the actual attribute value is):

````````
browser.div(data_tooltip: true).present?
#=> true

browser.div(data_tooltip: false).present?
#=> false
````````

In other words, `data_tooltip: true` and `data_tooltip: "true"` would mean two very different things. I do not think it adds any more confusion compared to the existing `data_tooltip: "true"` vs `data_tooltip: /true/`. The specifier value class has always had a meaning - String (exact match), Regexp (partial match) and now Boolean (presence).

**Comment by titusfortner** [comment]
just wanted to add my :+1: to @jkotests' proposed implementation.

**Comment by p0deje** [comment]
I have pushed quick and dirty support for attribute presence/absence selector. I haven't tested it extensively though so if anyone has a test suite to run with this commit, that would be great. Code review is appreciated as well.

**Comment by p0deje** [comment]
@titusfortner Care to take a quick look?

**Comment by titusfortner** [comment]
Apparently I never took a quick look. Do we still want to add this?

**Comment by p0deje** [comment]
Why not, we can easily add it if you think it makes sense.

**Comment by titusfortner** [comment]
yes, actually, the  more I look at it, the more I like this feature. We'll need to change it for the new locator separations, still, right?

**Comment by p0deje** [comment]
Yeah, I can do that.

**Comment by titusfortner** [comment]
out of curiosity, can this currently be done with: `browser.div(data_tooltip: /.*/).present?`

**Comment by p0deje** [comment]
Probably yes, though I haven't checked. Are you good to merge this feature?

**Comment by titusfortner** [comment]
My only hesitance is that it is multiple additional conditionals added to the code when the requirement could be satisfied with `/.+/` instead of `true`.

But this code does make it easier to understand I guess, so sure, let's go ahead and merge it.

**Comment by nonkor** [comment]
:tada:

**Comment by p0deje** [comment]
@nonkor It took us only 2 years to have this 🍿
